### ISSUE TITLE
Getting the same output logits regardless of input tensors

### ISSUE THREAD

**Issue by lolzz192** [Tencent/TurboTransformers#256]
Hello - I'm using AlbertModel with TurboTransformers 0.6 w/ CUDA 11.2 in Docker.

I've defined a SeqClass model as below:

```
class OptimizedAlbertForSequenceClassification:  # create a new class for speeding up
    def __init__(
            self, optimized_model, classifier
    ):
        self.optimized_model =  optimized_model
        self.classifier = classifier

    def __call__(
`            self, 
              input_ids,
              attention_mask=None,
              token_type_ids=None,
              position_ids=None,
              return_type=None,
      ):
        outputs = self.optimized_model(
            input_ids,
            attention_mask,
            token_type_ids,
            position_ids,
            return_type=return_type,
        )
        pooled_output = outputs[1]
        logits = self.classifier(
            pooled_output
        ) 
        return logits
```
A model is then initialized from a PyTorch Transformers model, all which runs without issue.

Inference is done using:

```
def make_pred(text):  
  tokens=tokenizer.encode(text, padding=True, truncation=True, max_length=256) # AlbertTokenizer
  tokens_tensor = torch.as_tensor([tokens], device=device) # device is cuda
  predicted_labels = torch.softmax(merged(tokens_tensor)[0], dim=-1) # merged is model defined above
  return predicted_labels.cpu().numpy()
```

The end result is:

["0.6927204", "0.3072796"], regardless of the input.

I can confirm the tokenizer works as expected, and the un-optimized model also is returns the proper predictions (switching merged() for the original, .from_pretrained() model).

Any ideas as to what's causing this?

Thank you!

**Comment by feifeibear** [comment]
Do you suggest that the end result is wrong?
Did you refer to the Albert example?
https://github.com/Tencent/TurboTransformers/blob/master/example/python/albert_example.py

**Comment by lolzz192** [comment]
> Do you suggest that the end result is wrong? Did you refer to the Albert example? https://github.com/Tencent/TurboTransformers/blob/master/example/python/albert_example.py

Yeah, as I'm getting exactly identical outputs regardless of input. For example: 

`make_pred("Hello!")`

> ["0.6927204", "0.3072796"]

`make_pred("What is that?")`

> ["0.6927204", "0.3072796"],

In contrast to the unconverted model:

`make_pred("Hello!")`

[0.00353016 0.99646986]

`make_pred("What is that?")`

[0.00238876 0.9976113 ]

What is rather odd is that running the exact same code in a different environment works properly (output logits match up). Any idea where I should look that might be causing this?

**Comment by feifeibear** [comment]
> > Do you suggest that the end result is wrong? Did you refer to the Albert example? https://github.com/Tencent/TurboTransformers/blob/master/example/python/albert_example.py
> 
> Yeah, as I'm getting exactly identical outputs regardless of input. For example:
> 
> `make_pred("Hello!")`
> 
> > ["0.6927204", "0.3072796"]
> 
> `make_pred("What is that?")`
> 
> > ["0.6927204", "0.3072796"],
> 
> In contrast to the unconverted model:
> 
> `make_pred("Hello!")`
> 
> [0.00353016 0.99646986]
> 
> `make_pred("What is that?")`
> 
> [0.00238876 0.9976113 ]
> 
> What is rather odd is that running the exact same code in a different environment works properly (output logits match up). Any idea where I should look that might be causing this?

**Comment by jk2625** [comment]
Sorry, I should have clarified. 

I built it on colab (bumped cmake to v22, pip'd MKL libraries/onnx) and ran the code above, and it worked correctly. 

Then I did something similar from the Nvidia Cuda 11.2-cudnn based image (installed python3.7, miniconda,  then conda install mkl mkl-include, pip'd transformers, torch, onnx, tf, a few other libraries, then git cloned and built TurboTransformers with cmake. Installs with no problem). Running the same code, I get the problem noted prior.

Happy to send any code/build files you need. 

Thank you!

**Comment by lolzz192** [comment]
Sorry, I should have clarified.

I built it on colab (bumped cmake to v22, pip'd MKL libraries/onnx) and ran the code above, and it worked correctly.

Then I did something similar from the Nvidia Cuda 11.2-cudnn based image (installed python3.7, miniconda, then conda install mkl mkl-include, pip'd transformers, torch, onnx, tf, a few other libraries, then git cloned and built TurboTransformers with cmake. Installs with no problem). Running the same code, I get the problem noted prior.

Happy to send any code/build files you need.

Thank you!

**Comment by feifeibear** [comment]
> 

It is very weird as you described. Please send me the test code and build files.

**Comment by lolzz192** [comment]
Of course. Where/how should I email/contact you?

**Comment by feifeibear** [comment]
Why not put your project in a GitHub repo? Please include a detailed doc to reproduce your experiments.

**Comment by lolzz192** [comment]
Here is the code and build files: https://github.com/lolzz192/TT-Torch-Docker

Thanks

**Comment by lolzz192** [comment]
Hello, sorry to reopen this, but I've done a bit more testing and it's still an issue.

With the new code provided, I'm getting the same problem - same output logits regardless of input on the converted model.

The identical code works on a T4 GPU, but not on a K80 GPU. I've make sure the driver versions + CUDA match exactly. 

Bert for Sequence Classification also fails on the K80:

`python3.7 bert_for_sequence_classification_example.py 
date       time         ( uptime  ) [ thread name/id ]                   file:line     v| 
2022-01-22 21:12:59.754 (   0.000s) [main thread     ]             loguru.cpp:610   INFO| arguments: turbo_transformers_cxx
2022-01-22 21:12:59.754 (   0.000s) [main thread     ]             loguru.cpp:613   INFO| Current dir: /home
2022-01-22 21:12:59.754 (   0.000s) [main thread     ]             loguru.cpp:615   INFO| stderr verbosity: 0
2022-01-22 21:12:59.754 (   0.000s) [main thread     ]             loguru.cpp:616   INFO| -----------------------------------
Some weights of the model checkpoint at bert-base-uncased were not used when initializing BertForSequenceClassification: ['cls.predictions.bias', 'cls.predictions.transform.dense.bias', 'cls.predictions.transform.LayerNorm.bias', 'cls.predictions.decoder.weight', 'cls.predictions.transform.dense.weight', 'cls.seq_relationship.bias', 'cls.predictions.transform.LayerNorm.weight', 'cls.seq_relationship.weight']
- This IS expected if you are initializing BertForSequenceClassification from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing BertForSequenceClassification from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).
Some weights of BertForSequenceClassification were not initialized from the model checkpoint at bert-base-uncased and are newly initialized: ['classifier.weight', 'classifier.bias']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.
Some weights of the model checkpoint at bert-base-uncased were not used when initializing BertForSequenceClassification: ['cls.predictions.bias', 'cls.predictions.transform.dense.bias', 'cls.predictions.transform.LayerNorm.bias', 'cls.predictions.decoder.weight', 'cls.predictions.transform.dense.weight', 'cls.seq_relationship.bias', 'cls.predictions.transform.LayerNorm.weight', 'cls.seq_relationship.weight']
- This IS expected if you are initializing BertForSequenceClassification from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing BertForSequenceClassification from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).
Some weights of BertForSequenceClassification were not initialized from the model checkpoint at bert-base-uncased and are newly initialized: ['classifier.weight', 'classifier.bias']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.
tensor([[0., 0.]], device='cuda:0', grad_fn=<AddmmBackward0>)
tensor([[-0.1572,  0.2813]], device='cuda:0', grad_fn=<AddmmBackward0>)
2022-01-22 21:13:09.408 (   9.654s) [main thread     ]             loguru.cpp:489   INFO| atexit
`

Thank you

**Comment by lolzz192** [comment]
Here is the output of nvidia-smi:

```
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 460.106.00   Driver Version: 460.106.00   CUDA Version: 11.2     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  Tesla K80           Off  | 00000000:00:04.0 Off |                    0 |
| N/A   46C    P0    61W / 149W |   2443MiB / 11441MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
                                                                               
+-----------------------------------------------------------------------------+
| Processes:                                                                  |
|  GPU   GI   CI        PID   Type   Process name                  GPU Memory |
|        ID   ID                                                   Usage      |
|=============================================================================|
|    0   N/A  N/A     18949      C   python3.7                        2440MiB |
+-----------------------------------------------------------------------------+
```
pip freeze:

```
absl-py==1.0.0
aiofiles==0.8.0
astunparse==1.6.3
beautifulsoup4==4.10.0
cached-property==1.5.2
cachetools==4.2.4
certifi==2021.10.8
charset-normalizer==2.0.10
click==8.0.3
ConfigArgParse==1.5.3
crcmod==1.7
distro-info==0.21
docopt==0.6.2
filelock==3.4.2
Flask==2.0.2
flatbuffers==2.0
future==0.18.2
gast==0.4.0
gdown==4.2.0
google-auth==2.3.3
google-auth-oauthlib==0.4.6
google-pasta==0.2.0
grpcio==1.43.0
h5py==3.6.0
httptools==0.3.0
huggingface-hub==0.4.0
idna==3.3
importlib-metadata==4.10.1
intel-openmp==2022.0.2
itsdangerous==2.0.1
Jinja2==3.0.3
joblib==1.1.0
keras==2.7.0
Keras-Preprocessing==1.1.2
libclang==12.0.0
Markdown==3.3.6
MarkupSafe==2.0.1
mkl==2022.0.2
mkl-include==2022.0.2
multidict==5.2.0
mysql-connector-python==8.0.28
ninja==1.10.2.3
numpy==1.21.5
oauthlib==3.1.1
onnx==1.10.2
onnxruntime-gpu==1.3.0
OpenNMT-py==1.1.0
opt-einsum==3.3.0
packaging==21.3
protobuf==3.19.3
pyasn1==0.4.8
pyasn1-modules==0.2.8
pycurl==7.43.0.2
PyGObject==3.30.4
pyonmttok==1.30.0
pyparsing==3.0.7
PySocks==1.7.1
python-apt==1.8.4.3
PyYAML==6.0
regex==2022.1.18
requests==2.27.1
requests-oauthlib==1.3.0
rsa==4.8
sacremoses==0.0.47
sanic==21.12.1
sanic-routing==0.7.2
six==1.16.0
soupsieve==2.3.1
tbb==2021.5.1
tensorboard==2.8.0
tensorboard-data-server==0.6.1
tensorboard-plugin-wit==1.8.1
tensorflow-estimator==2.7.0
tensorflow-gpu==2.7.0
tensorflow-io-gcs-filesystem==0.23.1
termcolor==1.1.0
tokenizers==0.10.3
torch==1.10.1
torchtext==0.4.0
tqdm==4.30.0
transformers==4.15.0
turbo-transformers @ file:///TurboTransformers/build/turbo_transformers/python/pypackage/dist/turbo_transformers-0.6.0-cp37-cp37m-linux_x86_64.whl
typing_extensions==4.0.1
ujson==5.1.0
unattended-upgrades==0.1
urllib3==1.26.8
uvloop==0.16.0
waitress==2.0.0
websockets==10.1
Werkzeug==2.0.2
wrapt==1.13.3
zipp==3.7.0
```

**Comment by feifeibear** [comment]
1. As mentioned by the warning. Some weights are not initialized correctly. It explains the wrong output of your application.
cls.predictions.bias', 'cls.predictions.transform.dense.bias', 'cls.predictions.transform.LayerNorm.bias', 'cls.predictions.decoder.weight', 'cls.predictions.transform.dense.weight', 'cls.seq_relationship.bias', 'cls.predictions.transform.LayerNorm.weight', 'cls.seq_relationship.weight

2. On K80, did you recompile the project? Because the GPU you using is too old, some arch options may not supports it.

**Comment by lolzz192** [comment]
> As mentioned by the warning. Some weights are not initialized correctly. It explains the wrong output of your application.
cls.predictions.bias', 'cls.predictions.transform.dense.bias', 'cls.predictions.transform.LayerNorm.bias', 'cls.predictions.decoder.weight', 'cls.predictions.transform.dense.weight', 'cls.seq_relationship.bias', 'cls.predictions.transform.LayerNorm.weight', 'cls.seq_relationship.weight

Those warnings happen on both GPUs, even the ones where it works properly. By default, random weights are used on model creation, so it should return some sort of varying output for each input.

2. Yes, I am re-compiling TT each time. I am using docker so each build builds everything. I'm building on the same machine running the code (running directly on the host doesn't work either.

**Comment by feifeibear** [comment]
> > As mentioned by the warning. Some weights are not initialized correctly. It explains the wrong output of your application.
> > cls.predictions.bias', 'cls.predictions.transform.dense.bias', 'cls.predictions.transform.LayerNorm.bias', 'cls.predictions.decoder.weight', 'cls.predictions.transform.dense.weight', 'cls.seq_relationship.bias', 'cls.predictions.transform.LayerNorm.weight', 'cls.seq_relationship.weight
> 
> Those warnings happen on both GPUs, even the ones where it works properly. By default, random weights are used on model creation, so it should return some sort of varying output for each input.
> 
> 2. Yes, I am re-compiling TT each time. I am using docker so each build builds everything. I'm building on the same machine running the code (running directly on the host doesn't work either.

Can we move the discussion to your specific codebase, if on github? I believe the problem you proposed is strongly related to your code and environment.

**Comment by lolzz192** [comment]
Ok, sure thing. Thanks
